### ISSUE TITLE
Cross referencing between documents

### ISSUE THREAD

**Issue by MartinHinz** [fiduswriter/fiduswriter#1118]
For editing books (e.g. monographs or anthologies) where the chapters are individual documents, it would be advantageous if there were a way to cross-reference headings, figures, tables or other anchors in another document.

**Comment by johanneswilm** [comment]
@MartinHinz Absolutely, this would be a great feature to have. The difficulty is in how to make such a feature work in the simple-most way that allows users to write all documents of a book at the same time while and allowing everyone to add/delete new entries at the same time and also allow those who organize the book to add/delete chapters.

**Comment by MartinHinz** [comment]
Thank you for your swift reply! Do you have any suggestions for a workaround? Much appreciated!

**Comment by nmueller18** [comment]
From a database perspective, this could only be an n:m-relationship with references to other instances (images, pages etc.) stored in a separate table. But is this feasible?

**Comment by johanneswilm** [comment]
Yes, I am not so concerned about being able to store it. It's more the
extra complications that come with webapps such as users usually having
slightly outdated versions of documents and books and how to react to that.
It doesn't make it an impossible feature. It just means one needs a thought
through plan on how to do it.

**Comment by johanneswilm** [comment]
So one way this could be done in a relatively simple manner - possibly: 

* We add the ability to add a "reference slug" to figures, tables, headings and in-text anchors. This is basically a free text field so you can put anything in there. for example "big crazy figure". If you just export that document itself, it will not export the reference title.

* In the link dialog, we add the option to create cross references to targets in other documents. Here you can also write anything, but it will auto-suggest to you all the titles that have recently been added to documents you have access to. If you export the document by itself, it will just use the "reference slug" as the title of the reference. So for example it could say "See *big crazy figure*"

* When creating a book, it will then try to link all the cross references within that book. If it cannot find them, it will do the same as when exporting a single document. If it can find more than one target with the same reference slug, it will chose the first one.

* On the book page, we add a "sanity check" It will load all the chapters as they exist in the database at that moment and run a number of checks on them, such as:

  * Are there targets for all cross references?
  * Are there more than one target for any given cross reference?
  * Are any tracked changes left in any of the chapters?
  * Are there unresolved comments in any of the chapters?
  * etc.

Such an approach should have an acceptable level of complexity to program and it should give you the flexibility to link between chapters without too high a learning curve. 

**Example:**
User A is participating in the writing of a book. B is the editor of the book. A has access to chapters  3 and 7. A wants to link to a figure in chapter 3 from chapter 7.

* First user A enters chapter 3. 
* A then finds the figure and clicks the menu button. 
* A then enters the reference title "Quantitative Easing Graph".
* A then exits chapter 3 and enters chapter 7.
* A clicks into the text at the place where the crossreference is to appear
* A then clicks the link button and chooses "Cross-reference to another document" as type. 
* As reference title, A starts typing "Quanti..."
* The system recognizes that this is a reference title that has recently been used in another document to which A has access and therefore suggests "Quantitative Easing Graph".
* A clicks on the suggestion. 
* At the place in the document where the crossreference is to appear, the text "Quantitative Easing Graph" appears with a grey background.
* B enters the book overview.
* B selects the book that contains both chapters and in the drop-down menu chooses "sanity check".
* The system finds the link from chapter 7 to 3 and sees that there is one target and one target only for the cross-reference in chapter 7. The sanity check, therefore, returns a positive result.
* B then chooses to export the book as an epub.
* As the system prepares the epub, it replaces the cross-reference text "Quantitative Easing Graph" with "Chapter 3, Figure 25". A link is placed as well.

How does that sound?

**Comment by MartinHinz** [comment]
From the first read, it sounds pretty awesome! Thanks for the effort! Does this also affect e.g. the JATS and the LaTeX export?
